cunumeric/module.py

```diff
@@ -2100,11 +2099,36 @@ def repeat(a, repeats, axis=None):
                 category=UserWarning,
             )
         repeats = np.int64(repeats)
-        result = array._thunk.repeat(
-            repeats=repeats,
-            axis=axis,
-            scalar_repeats=True,
-        )
+        if repeats < 0:
+            return ValueError(
+                "'repeats' should not be negative: {}".format(repeats)
+            )
+
+        # check output shape (if it will fit to GPU or not)
+        out_shape = list(array.shape)
+        out_shape[axis] *= repeats
+        out_shape = tuple(out_shape)
+        size = sum(out_shape) * array.itemsize
+        # check if size of the output array is less 8GB. In this case we can
+        # use output regions, otherwise we will use statcally allocated
+        # array
+        if size < 8589934592 / 2:
+
+            result = array._thunk.repeat(
+                repeats=repeats, axis=axis, scalar_repeats=True
+            )
+        else:
+            # this implementation is taken from CuPy
+            result = ndarray(shape=out_shape, dtype=array.dtype)
+            a_index = [slice(None)] * len(out_shape)
+            res_index = list(a_index)
+            offset = 0
+            for i in range(a._shape[axis]):
+                a_index[axis] = slice(i, i + 1)
+                res_index[axis] = slice(offset, offset + repeats)
+                result[res_index] = array[a_index]
+                offset += repeats
```

Comment:
I don't love this. We are emitting a separate operation for each slice. It would be more efficient if we could manually partition the `result` array on `axis` using a custom coloring, following a partitioning of `a` on the same axis:

```
p_a = axis_equal_partition(a, axis, colors)
for c in colors:
  result_coloring[c] = Rect(
    lo=p_a[c].domain.lo * repeats,
    hi=p_a[c].domain.hi * repeats
  )
p_result = partition_by_coloring(result, result_coloring)
for c in colors:
  repeat(p_result[c], p_a[c])  # index launch
```

However, this would require support for manual-coloring partitioning from the core. @magnatelee is this something that's reasonable?

Reply:
I don't think anything that does coloring would be any more scalable than the original code. and it is in some sense worse as it takes away the core's ability to reason about the partitioning. for example, if the core saw multiple tasks operating on  disjoint parts of the same array, it could potentially partition and map them in a way that those tasks are distributed in a balanced manner.